src/ILCompiler.Compiler/src/Compiler/TypePreinit.cs

```diff
@@ -0,0 +1,1964 @@
+                        }
+                        break;
+
+                    case ILOpcode.call:
+                        {
+                            MethodDesc method = (MethodDesc)methodIL.GetObject(reader.ReadILToken());
+                            MethodSignature methodSig = method.Signature;
+                            int paramOffset = methodSig.IsStatic ? 0 : 1;
+                            int numParams = methodSig.Length + paramOffset;
+
+                            TypeDesc owningType = method.OwningType;
+                            if (!_compilationGroup.CanInline(methodIL.OwningMethod, method)
+                                || (owningType.HasStaticConstructor
+                                    && owningType != methodIL.OwningMethod.OwningType
+                                    && !((MetadataType)owningType).IsBeforeFieldInit))
+                            {
+                                return Status.Fail(methodIL.OwningMethod, opcode);
+                            }
+
+                            Value[] methodParams = new Value[numParams];
+                            for (int i = numParams - 1; i >= 0; i--)
+                            {
+                                methodParams[i] = stack.PopIntoLocation(GetArgType(method, i));
+                            }
+
+                            Value retVal;
+                            if (!method.IsIntrinsic || !TryHandleIntrinsicCall(method, methodParams, out retVal))
+                            {
+                                recursionProtect ??= new Stack<MethodDesc>();
+                                recursionProtect.Push(methodIL.OwningMethod);
+                                Status callResult = TryScanMethod(method, methodParams, recursionProtect, out retVal);
+                                if (!callResult.IsSuccessful)
+                                {
+                                    recursionProtect.Pop();
+                                    return callResult;
+                                }
+                                recursionProtect.Pop();
+                            }
+
+                            if (!methodSig.ReturnType.IsVoid)
+                                stack.PushFromLocation(methodSig.ReturnType, retVal);
+                        }
+                        break;
+
+                    case ILOpcode.newobj:
+                        {
+                            MethodDesc ctor = (MethodDesc)methodIL.GetObject(reader.ReadILToken());
+                            MethodSignature ctorSig = ctor.Signature;
+
+                            TypeDesc owningType = ctor.OwningType;
+                            if (!_compilationGroup.CanInline(methodIL.OwningMethod, ctor)
+                                || !_compilationGroup.ContainsType(owningType)
+                                || !owningType.IsDefType
+                                || owningType.HasFinalizer
```

Comment:
> no finalizers - these never get garbage collected

A similar observable difference is also present without finalizers, e.g.:

```
using System;
using System.Runtime.CompilerServices;

class Program
{
    static Program s_instance = new Program();

    [MethodImpl(MethodImplOptions.NoInlining)]
    static WeakReference MakeWeakReference()
        => new WeakReference(s_instance);

    static void Main(string[] args)
    {
        var wk = MakeWeakReference();
        s_instance = new Program();
        Console.WriteLine(wk.IsAlive);
        GC.Collect();
        GC.WaitForPendingFinalizers();
        Console.WriteLine(wk.IsAlive);
    }
}
```

It prints "true, false" before this change and "true, true" after this change. I think it is fine to exclude finalizers here to make it less likely that this will introduce compat issues. 

Is `HasFinalizer` just a quirk for better compat, not required for runtime correctness? If it is the case, it may be worth adding a comment about it. 

If the finalizer condition was skipped here, the system would behave as if we had system with conservative GC and the conservative GC happened to not collect the object with the finalizer.

Reply:
I think it's just a compat quirk - I wanted to avoid situation when the finalizer would be side effecting. It probably doesn't matter in reality. Good point about WeakHandle too. I'll add a comment.